### PR TITLE
[RELAY] IR builder stablize refactor, clean pass

### PR DESCRIPTION
**Pull request by tqchen**
This is a clean refactor PR to attempt stablize the relay python IR building API, it also contains fixes and patches to enable text format. 

One of the general guidelines behind this refactor are:

- Minimum abstraction when possible (simplification of IRBuilder)
- Python API consistency (env.merge->env.update)
- Make ir_pass style consistent
   - Expr is always the first argument
   - Not require env to be present, when necessary, pass as a second argument

## List of Changes
- [x]  Simplify ir_builder to scope_builder cc @jroesch 
   -  ScopeBuilder is only used to build if and let scopes.
   - Use relay.Function to build functions (single way to build functions)
- [x] Rewrite all the test cases with the new style.
- [x] Add text print test to each op, fix problems that surface due to the new testcases.
- [x] Improve the BroadcastRel to handle symbolic case cc @jroesch 
- [x] Fix TupleWrapper cc @slyubomirsky 
    - To make sure python unpack works, we need to throw index error when there is out of bound access
    - TupleWrapper cannot assume the input field is a tuple type
- [x] Enable TulpleGetItem type inference with incomplete type
- [x] Fix dead code elimination when type is bool cc @MarisaKirisame 
- [x] Fix LeakyReluAttrs  cc @siju-samuel 
- [x] Merge reduce tests into a more concise functioncc @siju-samuel 
- [x] Rename TypeParam->TypeVar as suggested by @joshpoll

### CODE REVIEW

File: include/tvm/relay/attrs/nn.h [hunk 1]
```diff
@@ -254,7 +254,7 @@ struct PadAttrs : public tvm::AttrsNode<PadAttrs> {
 struct LeakyReluAttrs : public tvm::AttrsNode<LeakyReluAttrs> {
   double alpha;
 
-  TVM_DECLARE_ATTRS(DenseAttrs, "relay.attrs.LeakyReluAttrs") {
+  TVM_DECLARE_ATTRS(LeakyReluAttrs, "relay.attrs.LeakyReluAttrs") {
```
Comment (tqchen) [inline include/tvm/relay/attrs/nn.h]: @siju-samuel please be a bit more mindful next time

**Comment by junrushao1994** [comment]
Will do tonight

File: src/relay/ir/environment.cc [hunk 2]
```diff
@@ -16,71 +16,60 @@ using namespace runtime;
...
-GlobalVar EnvironmentNode::GetGlobalVar(const std::string &str) {
-  auto global_id = global_map_.find(str);
-  if (global_id != global_map_.end()) {
-    return (*global_id).second;
-  } else {
-    auto id = GlobalVarNode::make(str);
-    this->global_map_.Set(str, id);
-    return id;
-  }
+GlobalVar EnvironmentNode::GetGlobalVar(const std::string &name) {
```
Comment (MarisaKirisame) [inline src/relay/ir/environment.cc]: const std::string& name

File: src/relay/ir/environment.cc [hunk 3]
```diff
@@ -16,71 +16,60 @@ using namespace runtime;
...
-
-      this->functions.Set(var, checked_func);
-    } else {
-      this->functions.Set(var, checked_func);
+  Function checked_func = InferType(func, env, var);
+  auto type = checked_func->checked_type();
+  CHECK(type.as<IncompleteTypeNode>() == nullptr);
+  if (functions.find(var) != functions.end()) {
+    if (!update) {
+      throw dmlc::Error("already have definition for XXXX.");
```
Comment (MarisaKirisame) [inline src/relay/ir/environment.cc]: do XXXX

**Pull request merged**